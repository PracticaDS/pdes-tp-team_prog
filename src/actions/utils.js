export const BEGIN = 'BEGIN'
export const SUCCESS = 'SUCCESS'
export const ERROR = 'ERROR'
export const GET = 'GET'
export const POST = 'POST'
export const PUT = 'PUT'
export const DELETE = 'DELETE'

export const customFetch = (api, req) =>
  new Promise((resolve, reject) => {
    try {
      setTimeout(() => {
        resolve(req)
      }, 1000)
    } catch (err) {
      reject(err)
    }
  })

export const createUpdateGameRequest = game => ({
  method: 'PUT',
  mode: 'cors',
  headers: {
    'Content-Type': 'application/json',
  },
  body: JSON.stringify(game),
})

export const API_URL = process.env.API_URL || 'http://localhost:8080/'

export const createAction = (type, body) => ({ type, body })

const createRequest = (payload, method) => ({
  method,
  mode: 'cors',
  headers: {
    'Content-Type': 'application/json',
  },
  ...(method !== 'GET' && payload && { body: JSON.stringify(payload) }),
})

export const createThunk = ({
  action,
  url,
  payload,
  parse = ({ result }) => result,
  method,
}) => dispatch => {
  dispatch(createAction(`${action}_${BEGIN}`))
  return fetch(`${API_URL}${url}`, createRequest(payload, method))
    .then(result => result.json())
    .then(result => {
      dispatch(createAction(`${action}_${SUCCESS}`, parse({ result, type: SUCCESS })))
    })
    .catch(result => {
      dispatch(createAction(`${action}_${SUCCESS}`, parse({ result, type: ERROR })))
    })
}
