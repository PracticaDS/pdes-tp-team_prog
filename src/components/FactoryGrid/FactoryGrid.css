.FactoryGrid {
  display: flex;
  align-content: center;
  flex-direction: column;
  align-items: center;
}

.FactoryRow {
  display: flex;
  flex-direction: row;
}

.Bla {
  display: flex;
  justify-content: center;
  align-items: center;
}

.Material {
  z-index: 2;
  width: 1rem;
  height: 1rem;
}

.MachineImg {
  z-index: 1;
}
