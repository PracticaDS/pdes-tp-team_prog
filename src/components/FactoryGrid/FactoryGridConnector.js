import { connect } from 'react-redux'
import { selectDimensions } from '../../selectors/Grid'

const connector = FactoryGrid => {
  const mapStateToProps = state => ({
    dimensions: selectDimensions(state),
  })

  return connect(mapStateToProps)(FactoryGrid)
}

export default connector
