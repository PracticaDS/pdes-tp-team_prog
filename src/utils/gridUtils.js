import { createEmptyBlock } from './blockUtils'
import { createPosition } from './positionUtils'

export const generateEmptyGrid = (MAX_ROW, MAX_COLUMN) => {
  const result = new Array(MAX_ROW)
  for (let row = 0; row < MAX_ROW; row++) {
    result[row] = new Array(MAX_COLUMN)
    for (let column = 0; column < MAX_COLUMN; column++) {
      result[row][column] = createEmptyBlock({ position: createPosition(row, column) })
    }
  }
  return result
}

export const updateGridPosition = (grid, { row, column }, block) => {
  const newGrid = grid

  newGrid[row][column] = block

  return newGrid
}
