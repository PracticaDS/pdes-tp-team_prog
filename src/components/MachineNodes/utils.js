import { TIMER_TIME } from '../../utils/defaultValues'
import { isPositionValid } from '../../utils/directions'

const TIMEOUT_TIME = TIMER_TIME / 2

export const incrementCount = prevState => ({
  personalCount: prevState.personalCount + 1,
})

export const resetCount = () => ({
  personalCount: 0,
})

export const deactivate = () => ({
  isActive: false,
})

export const activate = () => ({
  isActive: true,
})

export const activateNode = () => ({
  ...resetCount(),
  ...activate(),
})

export const withTimeout = cb => {
  setTimeout(() => {
    cb()
  }, TIMEOUT_TIME)
}

export const areAllMaterials = (necessaryMaterials, currentMaterials) =>
  necessaryMaterials.every(materialRecipe => {
    const materialInGrid = currentMaterials[materialRecipe.material.name]
    return materialInGrid && materialInGrid.quantity >= materialRecipe.quantity
  })

export const materialsFiltered = (necessaryMaterials, currentMaterials) =>
  necessaryMaterials.map(materialRecipe => {
    const materialInGrid = currentMaterials[materialRecipe.material.name]
    return { ...materialInGrid, quantity: materialRecipe.quantity }
  })

export const hasNecessaryMaterials = (machine, items) => {
  const { metadata } = machine
  const { recipeSelected } = metadata
  const { necessaryMaterials } = recipeSelected

  return areAllMaterials(necessaryMaterials, items)
}

export const getItemsFromMaterials = (machine, items) => {
  const { metadata } = machine
  const { recipeSelected } = metadata
  const { necessaryMaterials } = recipeSelected

  return materialsFiltered(necessaryMaterials, items)
}

export const getCraftedMaterial = machine => {
  const { metadata } = machine
  const { recipeSelected } = metadata
  const { result } = recipeSelected
  const { material, quantity } = result

  return { material, quantity }
}

// machine - state
export const isFrequencyFinished = ({ node }, { personalCount }) =>
  node.machine && node.machine.frequency && node.machine.frequency <= personalCount

export const runCBInValidPosition = (position, dimensions, cb) => {
  if (isPositionValid(position, dimensions)) {
    cb()
  }
}
