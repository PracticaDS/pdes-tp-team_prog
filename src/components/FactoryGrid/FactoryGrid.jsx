import React from 'react'
import './FactoryGrid.css'
import connector from './FactoryGridConnector'
import Block from '../FactoryBlock/BlockConnected'
import { machineByType } from '../../utils/machineUtils'

const simpleArray = (n, m) => Array(m).fill(Array(n).fill(''))

const renderBlock = (position, renderNode) => (
  <Block key={`${position.row}_${position.column}`} position={position} renderNode={renderNode} />
)

const renderNode = node => {
  const MachineNode = machineByType[node.machine.type]
  return <MachineNode className="MachineNode" node={node} />
}

const FactoryGrid = ({ dimensions: { n, m } }) => (
  <div className="FactoryGrid">
    {simpleArray(n, m).map((RowComponent, Rindex) => (
      <div key={Rindex} className="FactoryRow">
        {RowComponent.map((node, CIndex) =>
          renderBlock({ row: Rindex, column: CIndex }, renderNode),
        )}
      </div>
    ))}
  </div>
)

export default connector(FactoryGrid)
