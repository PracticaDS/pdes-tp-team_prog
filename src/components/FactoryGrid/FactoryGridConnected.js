import React from 'react'
import FactoryGrid from './FactoryGrid'
import FactoryConnector from './FactoryGridConnector'
import { machineByType } from '../../utils/machineUtils'
import Block from '../FactoryBlock/BlockConnected'

const renderNode = node => {
  const MachineNode = machineByType[node.machine.type]
  return <MachineNode className="MachineNode" node={node} />
}

const renderBlock = position => (
  <Block key={`${position.row}_${position.column}`} position={position} renderNode={renderNode} />
)

const factoryGrid = props => <FactoryGrid {...props} renderBlock={renderBlock} />

export default FactoryConnector(factoryGrid)
