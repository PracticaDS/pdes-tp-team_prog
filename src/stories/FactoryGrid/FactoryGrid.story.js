/* eslint-disable react/jsx-key */
/* eslint-disable react/jsx-filename-extension */
import React from 'react'
import { storiesOf } from '@storybook/react'
import FactoryGrid from '../../components/FactoryGrid/FactoryGrid'
import CrafterMachineNode from '../../components/MachineNodes/CrafterMachineNode/CrafterMachineNode'
import FurnaceMachineNode from '../../components/MachineNodes/FurnaceMachineNode/FurnaceMachineNode'
import SellerMachineNode from '../../components/MachineNodes/SellerMachineNode/SellerMachineNode'
import TransporterMachineNode from '../../components/MachineNodes/TransporterMachineNode/TransporterMachineNode'
import EmptyMachineNode from '../../components/MachineNodes/EmptyMachineNode/EmptyMachineNode'
import StarterMachineNode from '../../components/MachineNodes/StarterMachineNode/StarterMachineNode'

storiesOf('FactoryGrid', module)
  .add('Simple Grid with 1x2 empty blocks', () => {
    const rowBlock = [EmptyMachineNode, EmptyMachineNode]
    const blocks = [rowBlock]
    return <FactoryGrid componenets={blocks} />
  })
  .add('Simple Grid with 2x1 empty blocks', () => {
    const rowBlock = [EmptyMachineNode]
    const blocks = [rowBlock, rowBlock]
    return <FactoryGrid componenets={blocks} />
  })
  .add('Simple Grid with 2x2 empty blocks', () => {
    const rowBlock = [EmptyMachineNode, EmptyMachineNode]
    const blocks = [rowBlock, rowBlock]
    return <FactoryGrid componenets={blocks} />
  })
  .add('Simple Grid with 5x4 empty blocks', () => {
    const rowBlock = [EmptyMachineNode, EmptyMachineNode, EmptyMachineNode, EmptyMachineNode]
    const blocks = [rowBlock, rowBlock, rowBlock, rowBlock, rowBlock]
    return <FactoryGrid componenets={blocks} />
  })
  .add('Grid with 6x6 and differents types of Machines', () => {
    const rowBlock1 = [
      EmptyMachineNode,
      CrafterMachineNode,
      EmptyMachineNode,
      SellerMachineNode,
      EmptyMachineNode,
      CrafterMachineNode,
    ]
    const rowBlock2 = [
      FurnaceMachineNode,
      EmptyMachineNode,
      EmptyMachineNode,
      EmptyMachineNode,
      EmptyMachineNode,
      EmptyMachineNode,
    ]
    const rowBlock3 = [
      TransporterMachineNode,
      EmptyMachineNode,
      CrafterMachineNode,
      EmptyMachineNode,
      TransporterMachineNode,
      EmptyMachineNode,
    ]
    const rowBlock4 = [
      EmptyMachineNode,
      SellerMachineNode,
      EmptyMachineNode,
      CrafterMachineNode,
      FurnaceMachineNode,
      StarterMachineNode,
    ]
    const rowBlock5 = [
      FurnaceMachineNode,
      FurnaceMachineNode,
      EmptyMachineNode,
      EmptyMachineNode,
      EmptyMachineNode,
      CrafterMachineNode,
    ]
    const rowBlock6 = [
      TransporterMachineNode,
      EmptyMachineNode,
      CrafterMachineNode,
      StarterMachineNode,
      SellerMachineNode,
      EmptyMachineNode,
    ]
    const blocks = [rowBlock1, rowBlock2, rowBlock3, rowBlock4, rowBlock5, rowBlock6]
    return <FactoryGrid componenets={blocks} />
  })
