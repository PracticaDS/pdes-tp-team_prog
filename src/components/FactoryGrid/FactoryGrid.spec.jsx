import React from 'react'
import { mount } from 'enzyme'
import FactoryGrid from './FactoryGrid'
import { generateEmptyGrid } from '../../utils/gridUtils'

describe('Factory GridSuite', () => {
  describe('Basic rendering', () => {
    let dimensionsMock = {}
    let renderBlockMock = null
    let customGrid = []
    let FactoryGridComponent = null

    beforeEach(() => {
      dimensionsMock = { n: 2, m: 2 }
      renderBlockMock = jest.fn()
      customGrid = generateEmptyGrid(2, 2)
      FactoryGridComponent = mount(
        <FactoryGrid dimensions={dimensionsMock} renderBlock={renderBlockMock} />,
      )
    })

    it('render simple Factory Block component', () => {
      expect(FactoryGridComponent).toBeDefined()
    })
    it('render block should be called with every row and column object', () => {
      customGrid.forEach((row, rowIndex) => {
        row.forEach((column, columnIndex) => {
          expect(renderBlockMock).toBeCalledWith({ row: rowIndex, column: columnIndex })
        })
      })
    })
  })
})
