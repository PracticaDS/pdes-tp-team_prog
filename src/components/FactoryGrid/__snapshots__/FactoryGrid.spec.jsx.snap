// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render simple Grid Component with 2x2 Blocks 1`] = `
<div
  className="FactoryGrid"
>
  <div
    className="FactoryRow"
  >
    <div
      className="Block"
    >
      <div
        className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
      >
        <div
          className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
        />
      </div>
    </div>
    <div
      className="Block"
    >
      <div
        className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
      >
        <div
          className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
        />
      </div>
    </div>
  </div>
  <div
    className="FactoryRow"
  >
    <div
      className="Block"
    >
      <div
        className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
      >
        <div
          className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
        />
      </div>
    </div>
    <div
      className="Block"
    >
      <div
        className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
      >
        <div
          className="MuiPaper-root-2 MuiPaper-elevation1-5 MuiPaper-rounded-3 MuiCard-root-1"
        />
      </div>
    </div>
  </div>
</div>
`;
